src/irc/core/irc-cap.c

```diff
@@ -0,0 +1,241 @@
+#include "module.h"
+#include "signals.h"
+#include "misc.h"
+
+#include "irc-cap.h"
+#include "irc-servers.h"
+
+int cap_toggle (IRC_SERVER_REC *server, char *cap, int enable)
+{
+	/* This is meant to be used after the inital negotiation is done */
+	if (!server->cap_complete)
+		return FALSE;
+
+	if (enable && gslist_find_string(server->cap_supported, cap)) {
+		irc_send_cmdv(server, "CAP REQ %s", cap);
+		return TRUE;
+	}
+	else if (!enable && gslist_find_string(server->cap_active, cap)) {
+		irc_send_cmdv(server, "CAP REQ -%s", cap);
+		return TRUE;
+	}
+
+	return FALSE;
+}
+
+void cap_finish_negotiation (IRC_SERVER_REC *server)
+{
+	if (server->cap_complete)
+		return;
+
+	server->cap_complete = TRUE;
+	irc_send_cmd_now(server, "CAP END");
+
+	signal_emit("server cap end", 1, server);
+}
+
+static void cap_emit_signal (IRC_SERVER_REC *server, char *cmd, char *args)
+{
+	char *signal_name;
+
+	signal_name = g_strdup_printf("server cap %s %s", cmd, args? args: "");
+	signal_emit(signal_name, 1, server);
+	g_free(signal_name);
+}
+
+static void event_cap (IRC_SERVER_REC *server, char *args, char *nick, char *address)
+{
+	GSList *tmp;
+	GString *cmd;
+	char *params, *evt, *list, **caps;
+	int i, caps_length, disable;
+
+	params = event_get_params(args, 3, NULL, &evt, &list);
+	if (params == NULL)
+		return;
+
+	caps = g_strsplit(list, " ", -1);
+	caps_length = g_strv_length(caps);
+
+	if (!g_strcmp0(evt, "LS")) {
+		/* Create a list of the supported caps */
+		for (i = 0; i < caps_length; i++)
+			server->cap_supported = g_slist_append(server->cap_supported, g_strdup(caps[i]));
```

Comment:
>Note that `g_slist_append()` has to traverse the entire list to find the end, which is inefficient when adding multiple elements. A common idiom to avoid the inefficiency is to prepend the elements and reverse the list when all elements have been added

Since order doesn't really matter, replacing these calls with `g_slist_prepend()` should work.